pageserver/src/lib.rs

```diff
@@ -1,24 +1,59 @@
 use std::net::SocketAddr;
-use std::path::PathBuf;
 
 pub mod page_cache;
 pub mod page_service;
 pub mod pg_constants;
-pub mod restore_datadir;
-pub mod restore_s3;
+pub mod restore_local_repo;
 pub mod tui;
 pub mod tui_event;
 mod tui_logger;
 pub mod waldecoder;
 pub mod walreceiver;
 pub mod walredo;
+pub mod basebackup;
 
 #[derive(Debug, Clone)]
 pub struct PageServerConf {
-    pub data_dir: PathBuf,
     pub daemonize: bool,
     pub interactive: bool,
-    pub wal_producer_connstr: Option<String>,
     pub listen_addr: SocketAddr,
-    pub restore_from: String,
+}
+
+// Zenith Timeline ID is a 32-byte random ID.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct ZTimelineId([u8; 16]);
+
+impl ZTimelineId {
+
+    pub fn from_str(s: &str) -> Result<ZTimelineId, hex::FromHexError> {
+        let timelineid = hex::decode(s)?;
+
+        let mut buf: [u8; 16] = [0u8; 16];
+        buf.copy_from_slice(timelineid.as_slice());
+        Ok(ZTimelineId(buf))
+    }
+
+    pub fn from(b: [u8; 16]) -> ZTimelineId {
+        ZTimelineId(b)
+    }
+
+    pub fn get_from_buf(buf: &mut dyn bytes::Buf) -> ZTimelineId {
+        let mut arr = [0u8; 16];
+        buf.copy_to_slice(&mut arr);
+        ZTimelineId::from(arr)
+    }
+
+    pub fn as_arr(&self) -> [u8; 16] {
+        self.0
+    }
+
+    pub fn to_str(self: &ZTimelineId) -> String {
+         hex::encode(self.0)
+    }
```

Comment:
I would suggest this type `impl std::fmt::Display` instead; then you get a blanket impl of the `ToDisplay` trait, which includes the `to_string()` function. This is a better name anyway; `to_str` implies a `&str` return type.

Also, `Display` means you can do `println!("{}", timeline)` without calling anything explicitly to do the conversion.